gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/PubSub.java

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.pubsub;
+
+import com.google.cloud.AsyncPage;
+import com.google.cloud.Page;
+import com.google.cloud.Service;
+
+import java.io.Serializable;
+import java.util.List;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * An interface for Google Cloud Pub/Sub.
+ *
+ * @see <a href="https://cloud.google.com/pubsub/">Google Cloud Pub/Sub</a>
+ */
+public interface PubSub extends Service<PubSubOptions> {
```

Comment:
> I don't understand the need for the *Option classes. Why aren't those POJOs? How does it make the library easier to use?

Across the library we use the options as "optional settings for a request" which typically a user don't care about.
In that case a var-arg is nice as it does not force the user to deal with it unless desired.

Examples for that are: the possible set of fields to return, Storage advanced state matching (not-exists, already-exists).

Settings which are mandatory are not considered as an option and are part of the method signature.

This way lets the user do: `pubub.listSubscriptions()` for the common use case (especially as the return type
allows to handle the pages through it) and if desired `pubsub.listSubscription(ListOption.pageToken("bla"))`
when a user explicitly wants to start from a page.

Note that `pageToken` could be statically imported to make the call more concise.

An alternative for that would have been to create an immutable Option class with a Builder and require the user to supply it or to overload each method to be with or without it. We chose not to go this route.

> We opted for a single service class for consistency with all other modules which expose their functionality through a single interface.

I agree with @mziccard. That is consistent with what we have so far. Also, PubSub is represented to the user as one service. We provide a common way to create a service instance (associated with credentials, retry settings,...) and
I think such settings applies to both publisher and subscriber.

I think what you have here is a case of one service with 2 endpoints, each endpoint support a distinct set of functionality. For that we could make the PubSub service returns Publisher and Subscriber but I am not convinced
that this indirection would be helpful. Another option is to make PubSub implements 2 interfaces for Publisher and Subscriber APIs but I doubt how useful that is going to be given that we already provide the functional Subscription and Topic classes.

As a side note, I actually find this separation of the 2 services/end-points a bit artificial. Example for that is that I can see a need for subscribers to interact with topic (list or even create).